### PR TITLE
Writing tutorial for `deep reinforcement learning implementation with pytorch from scratch`

### PR DESCRIPTION
**Pull request by hahas94**
# Description

This PR is for a tutorial I have been working for, namely the `deep reinforcement learning implementation with pytorch from scratch`. There are no changes in other code/files, hence this RP only adds a new file, along with images and gifs. 

Fixes # (issue)

### Screenshots

# Checklist:

- [x] I have run the [`pre-commit` checks](https://pre-commit.com/) with `pre-commit run --all-files` (see `CONTRIBUTING.md` instructions to set it up)
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

**Comment by RedTachyon** [comment]
Was this tutorial discussed with anyone, on Discord or otherwise?

**Comment by hahas94** [comment]
Yes, see Issues `Tutorials #28`, [link](https://github.com/Farama-Foundation/Gymnasium/issues/28).

**Comment by pseudo-rnd-thoughts** [comment]
Thanks for the tutorial @hahas94, is this a new tutorial or has this been published before? 

Having a quick look it is very impressive, I haven't checked the implementation

My initial reaction is a couple of the more advanced points shouldn't be included, i.e., the `parallel_training`
A quick formatting point is that for code then you need to use two backticks not one (otherwise its italicized)

I'll try to make a more complete review this week

**Comment by hahas94** [comment]
Thank you!

No it has not been published before, so it is new.

Yes please have a thorough look and let me know what changes to make or how to restructure it if necessary.

File: docs/tutorials/training_agents/deep_rl_tutorial.py [hunk 1]
```diff
@@ -0,0 +1,1069 @@
...
+        self.episode_lengths[self.index] = episode_length
+        self.episode_action_values[self.index] = episode_action_value
+        self.policy_entropy[self.index] = entropy
+        self.losses[self.index] = loss
+
+        self.index += 1
+
+    def add_evaluation_return(self, mean_eval_return: float):
+        """Add mean evaluation return obtained at each evaluation point."""
+        self.evaluation_returns[self.eval_index] = mean_eval_return
```
[the remark this answers is not in the document]
Comment (hahas94) [reply docs/tutorials/training_agents/deep_rl_tutorial.py:181]: It is an array, not list.

File: docs/tutorials/training_agents/deep_rl_tutorial.py [hunk 2]
```diff
@@ -0,0 +1,1069 @@
...
+# visualization. Next, we need a function that takes an iterable or array as input and returns a tensor out of it,
+# which can be used to transform observations into torch tensors. We also have a namedtuple for agent's action info.
+# Lastly, we need a function to seed all sources of randomness, so that we can reproduce results for debugging,
+# but also show that the training is stable across a number of different seeds. Note however that when training
+# on `torch.cuda`, there will be some randomness as the things are seeded now, because the choice of algorithm for
+# a certain task can be different across runs, even on the same seed and hardware. Although it can be fixed, it
+# comes at cost on performance, therefore it is left to be random.
+#
+@dataclasses.dataclass
+class Hyperparameters:
```
[the remark this answers is not in the document]
Comment (hahas94) [reply docs/tutorials/training_agents/deep_rl_tutorial.py:72]: Done in `141f27e`

File: docs/tutorials/training_agents/deep_rl_tutorial.py [hunk 3]
```diff
@@ -0,0 +1,1069 @@
...
+    Returns:
+        tensor
+    """
+    tensor = torch.tensor(np.array(array), device=device)
+
+    if new_axis:
+        tensor = tensor.unsqueeze(0)
+
+    if normalize:
+        tensor = tensor / 255.0
```
[the remark this answers is not in the document]
Comment (hahas94) [reply docs/tutorials/training_agents/deep_rl_tutorial.py]: Done in `141f27e`

File: docs/tutorials/training_agents/deep_rl_tutorial.py [hunk 4]
```diff
@@ -0,0 +1,1069 @@
...
+# an environment, and wrap it with the necessary wrappers, such `AtariPreprocessing` for atari environments. The
+# `FrameStack` wrapper is used for all environments, even though it is not necessary for environments like Cart Pole,
+# however, in these cases only one observation is used. For some atari environments, an agent can receive a
+# termination signal once a life is lost, however this is not recommended (Machado et al. (2018)), therefore here
+# termination signal is issued only when a game is over.
+#
+def create_env(params: Hyperparameters, record_video: bool = False) -> gymnasium.Env:
+    """
+    Create an environment and apply AtariProcessing wrappers if it is an Atari environment.
+    Only environments with discrete action spaces are supported.
```
[the remark this answers is not in the document]
Comment (hahas94) [reply docs/tutorials/training_agents/deep_rl_tutorial.py]: Done in `141f27e`. Moved it to `train` function instead.

File: docs/tutorials/training_agents/deep_rl_tutorial.py [hunk 5]
```diff
@@ -0,0 +1,1069 @@
...
+
+# %%
+# 2. Environment
+# --------------
+# If we wanted to create environments, then its class implementation would go here. However, since we are going to
+# use `gymnasium` environments, all we need is a function to create them. The below function can be used to create
+# an environment, and wrap it with the necessary wrappers, such `AtariPreprocessing` for atari environments. The
+# `FrameStack` wrapper is used for all environments, even though it is not necessary for environments like Cart Pole,
+# however, in these cases only one observation is used. For some atari environments, an agent can receive a
+# termination signal once a life is lost, however this is not recommended (Machado et al. (2018)), therefore here
```
[the remark this answers is not in the document]
Comment (hahas94) [reply docs/tutorials/training_agents/deep_rl_tutorial.py]: Done in `141f27e`

File: docs/tutorials/training_agents/deep_rl_tutorial.py [hunk 6]
```diff
@@ -0,0 +1,1069 @@
...
+        self._params = params
+        self._buffer = collections.deque([], maxlen=params.capacity)
+
+    def push(
+        self,
+        obs: gymnasium.wrappers.LazyFrames,
+        action: int,
+        reward: gymnasium.core.SupportsFloat,
+        next_obs: gymnasium.wrappers.LazyFrames,
+        done: bool,
```
[the remark this answers is not in the document]
Comment (hahas94) [reply docs/tutorials/training_agents/deep_rl_tutorial.py]: Done in `141f27e`

File: docs/tutorials/training_agents/deep_rl_tutorial.py [hunk 7]
```diff
@@ -0,0 +1,1069 @@
...
+        Initialize the replay memory.
+        Args:
+            params: Hyperparameters
+        """
+        self._params = params
+        self._buffer = collections.deque([], maxlen=params.capacity)
+
+    def push(
+        self,
+        obs: gymnasium.wrappers.LazyFrames,
```
[the remark this answers is not in the document]
Comment (hahas94) [reply docs/tutorials/training_agents/deep_rl_tutorial.py]: Done in `141f27e`

File: docs/tutorials/training_agents/deep_rl_tutorial.py [hunk 8]
```diff
@@ -0,0 +1,1069 @@
...
+        out_features = self._params.n_actions * self._params.n_atoms
+        n_hidden_units = self._params.n_hidden_units
+
+        self._head = torch.nn.Sequential(
+            nn.Linear(in_features=in_features, out_features=n_hidden_units),
+            nn.ReLU(),
+            nn.Linear(in_features=n_hidden_units, out_features=out_features),
+        )
+
+    def _output_size(self) -> int:
```
[the remark this answers is not in the document]
Comment (hahas94) [reply docs/tutorials/training_agents/deep_rl_tutorial.py:379]: I am not sure how it could be computed from only `env.action_space.shape` or `env.observation_space.shape` without going through the steps made by the `Conv2d` layer, that is `kerner_size`, `stride` `padding` etc.

File: docs/tutorials/training_agents/deep_rl_tutorial.py [hunk 9]
```diff
@@ -0,0 +1,1069 @@
...
+# they become old, as well as providing a way to sample experiences randomly. The agent network defines the structure
+# of the neural network and a way to make predictions. The below classes implement exactly these things, that is an
+# ``Agent``, together with a ``ReplayMemory`` and a ``Network`` class. Because C51 uses two networks, the agent has
+# a *main* and a *target* network, and a way to ``update`` the target network's parameters to be equal to the
+# main network's.
+# Lastly, since an agent needs to do exploration using epsilon-greedy, there is a way for updating the exploration rate,
+# i.e. ``updating epsilon``.
+#
+class ReplayMemory:
+    """Implements a circular replay memory object based on a deque."""
```
[the remark this answers is not in the document]
Comment (hahas94) [reply docs/tutorials/training_agents/deep_rl_tutorial.py]: Goog point. `collections.deque` was used due to its simplicity. I actually had done profiling to compare `collections.deque` with `list`, however on small scale, which turned out it didn't show the true difference in performance. Now I did it again, this time for 1M capacity, and indeed, `list` performs better in terms of time, slightly better memory-wise. 

I don't know what `ERB` is, maybe `Episodic Replay Buffer`?!

File: docs/tutorials/training_agents/deep_rl_tutorial.py [hunk 10]
```diff
@@ -0,0 +1,1069 @@
...
+        if self._epsilon > self._params.epsilon_end:
+            self._epsilon -= self._epsilon_decay
+
+    def update_target_network(self):
+        """Updating the parameters of the target network to equal the main network's parameters."""
+        self._target_network.load_state_dict(self._main_network.state_dict())
+
+    def learn(self) -> float:
+        """Learning step, updates the main network through backpropagation. Returns loss."""
+        obs, actions, rewards, next_obs, dones = self.replay_memory.sample()
```
[the remark this answers is not in the document]
Comment (hahas94) [reply docs/tutorials/training_agents/deep_rl_tutorial.py]: Done in `141f27e`

File: docs/tutorials/training_agents/deep_rl_tutorial.py [hunk 11]
```diff
@@ -0,0 +1,1069 @@
...
+
+
+# %%
+# 4. Training and evaluation
+# --------------------------
+# In order to train an agent on a specific environment, we need a function to implement the interaction between the
+# two. The ``train`` function below does that, where it creates an agent and an environment based on provided
+# hyperparameters, creates buffers for storing intermediate results, and runs the training process. It also stores
+# the results into a ``MetricsLogger`` object periodically, and once training has finished this logger is returned.
+# We also need to evaluate the agent periodically, and a standard proposed by Machado et al. (2018), an agent should
```
[the remark this answers is not in the document]
Comment (hahas94) [reply docs/tutorials/training_agents/deep_rl_tutorial.py]: Meant to say they propose it to become a standard. Have reworded it now.

File: docs/tutorials/training_agents/deep_rl_tutorial.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (Kallinteris-Andreas) [reply docs/tutorials/training_agents/deep_rl_tutorial.py:379]: doesn't `output_size == action_space.shape[0]`??

[the remark this answers is not in the document]
Comment (hahas94) [reply docs/tutorials/training_agents/deep_rl_tutorial.py:379]: If we are talking about the output of the network, then in C51 it is `action_space.shape[0] * n_atoms`.

However, note the two lines 
```
in_features = self._output_size()
out_features = self._params.n_actions * self._params.n_atoms
```
where the method `_output_size` is used to compute the output size of the flatten layer. Maybe the choice of method name is not good?

File: docs/tutorials/training_agents/deep_rl_tutorial.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (Kallinteris-Andreas) [reply docs/tutorials/training_agents/deep_rl_tutorial.py]: ERB means "Experience Replay Buffer"

File: docs/tutorials/training_agents/deep_rl_tutorial.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (Kallinteris-Andreas) [reply docs/tutorials/training_agents/deep_rl_tutorial.py:379]: "`in_features = self._output_size()`" I very confused by what I am reading here, maybe you need to rename the function `_output_size` to something else, or add some more comments

File: docs/tutorials/training_agents/deep_rl_tutorial.py [hunk 12]
```diff
@@ -0,0 +1,1095 @@
...
+    fig, axes = plt.subplots(nrows=3, ncols=2, figsize=(14, 21))
+
+    eval_axes = axes.flatten()[-1]
+    axes = axes.flatten()[:-1]
+
+    for i, ax in enumerate(axes):
+        mean, std = processed_data[i]
+        ax.plot(x, mean, color=color, label="mean")
+        ax.fill_between(
+            x=x, y1=mean - std, y2=mean + std, label="stddev", alpha=0.2, color="tomato"
```
Comment (Kallinteris-Andreas) [inline docs/tutorials/training_agents/deep_rl_tutorial.py:882]: Since when is "tomato" a color? lol

Does `matplotlib` not complain???

File: docs/tutorials/training_agents/deep_rl_tutorial.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (hahas94) [reply docs/tutorials/training_agents/deep_rl_tutorial.py]: Suspected that. Changed the method name to:
```    
def _fc_input_size(self) -> int:
        """Compute size of input to first linear layer."""
```

File: docs/tutorials/training_agents/deep_rl_tutorial.py [hunk 13]
```diff
@@ -0,0 +1,1095 @@
...
+# comes at cost on performance, therefore it is left to be random.
+#
+@dataclasses.dataclass
+class Hyperparameters:
+    """Hyperparameters class for Categorical-DQN."""
+
+    # --- env related ---
+    env_name: str = ""
+    obs_shape: tuple = (1,)  # shape of observation, ex. (1, 4), (84, 84)
+    n_actions: int = 0
```
[the remark this answers is not in the document]
Comment (hahas94) [reply docs/tutorials/training_agents/deep_rl_tutorial.py:78]: Not only these two, but several other parameters are in this class that are not hyperparameters, however this way it is easier to pass these parameters to any function/class that need them.

File: docs/tutorials/training_agents/deep_rl_tutorial.py (hunk 12 above)
Comment (hahas94) [reply docs/tutorials/training_agents/deep_rl_tutorial.py:882]: :D it is a beautiful one too. See [List of named colors](https://matplotlib.org/stable/gallery/color/named_colors.html)

File: docs/tutorials/training_agents/deep_rl_tutorial.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (Kallinteris-Andreas) [reply docs/tutorials/training_agents/deep_rl_tutorial.py]: isn't the case that:
`in_features = env.observation_space.shape[0]`

[the remark this answers is not in the document]
Comment (hahas94) [reply docs/tutorials/training_agents/deep_rl_tutorial.py]: Partially true.

If the vision part of the network is not utilized, then indeed `in_features = env.observation_space.shape[0]`, and this is exactly what is returned by `_fc_input_size`, since the convolutional sequential is empty, so the observation is returned as is.

But if the vision part is active, then it depends on the observation shape and convolutional operations.
Assuming that `env.observation_space.shape == (4, 84, 84)` and the current convolutional structure, then the size is computed manually as follows:
the output shape of the first conv2d layer is `(32, 20, 20)`,
the output shape of the second conv2d layer is `(64, 9, 9)` and 
the output shape of the third conv2d layer is `(64, 7, 7)`
which will produce `64 * 7 * 7 = 3136` weights at the flatten layer.

But we can't always assume that the `env.observation_space.shape == (4, 84, 84)` or that the network has this structure in general. So I prefer computing it using the current method.

**Comment by hahas94** [comment]
@pseudo-rnd-thoughts, @Kallinteris-Andreas 

just checking in on this PR. Is there anything more I need to address or any additional feedback you can provide?

**Comment by JDRanpariya** [comment]
Is there a reason this PR is not merged but closed?
